source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2052 @@
+
+        args->nw_socket = s_socket_acquire_internal_ref(nw_socket);
+        args->allocator = socket->allocator;
+        args->error_code = error_code;
+
+        aws_task_init(&args->task, s_process_connection_result_task, args, "connectionSuccessTask");
+        aws_event_loop_schedule_task_now(nw_socket->event_loop, &args->task);
+    }
+
+    aws_mutex_unlock(&nw_socket->synced_data.lock);
+}
+
+struct connection_state_change_args {
+    struct aws_task task;
+    struct aws_allocator *allocator;
+    struct aws_socket *socket;
+    struct nw_socket *nw_socket;
+    nw_connection_t nw_connection;
+    nw_connection_state_t state;
+    int error;
+};
+
+static void s_process_connection_state_changed_task(struct aws_task *task, void *args, enum aws_task_status status) {
+    (void)status;
+    (void)task;
+
+    struct connection_state_change_args *connection_args = args;
+
+    struct nw_socket *nw_socket = connection_args->nw_socket;
+    nw_connection_t nw_connection = connection_args->nw_connection;
+    nw_connection_state_t state = connection_args->state;
+
+    if (state == nw_connection_state_cancelled) {
+        AWS_LOGF_INFO(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: Apple network framework socket connection state changed to "
+            "nw_connection_state_cancelled. ",
+            (void *)nw_socket,
+            (void *)nw_socket->os_handle.nw_connection);
+        aws_mutex_lock(&nw_socket->synced_data.lock);
+        struct aws_socket *socket = nw_socket->synced_data.base_socket;
+        s_set_socket_state(nw_socket, socket, CLOSED);
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+
+        s_socket_release_internal_ref(nw_socket);
+
+    } else if (state == nw_connection_state_ready) { /* we're connected! */
+        AWS_LOGF_INFO(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: Apple network framework socket connection state changed to nw_connection_state_ready. ",
+            (void *)nw_socket,
+            (void *)nw_socket->os_handle.nw_connection);
+        nw_path_t path = nw_connection_copy_current_path(nw_connection);
+        nw_endpoint_t local_endpoint = nw_path_copy_effective_local_endpoint(path);
+        nw_release(path);
+        const char *hostname = nw_endpoint_get_hostname(local_endpoint);
+        uint16_t port = nw_endpoint_get_port(local_endpoint);
+
+        aws_mutex_lock(&nw_socket->synced_data.lock);
+        struct aws_socket *socket = nw_socket->synced_data.base_socket;
+        if (socket) {
+            if (hostname != NULL) {
+                size_t hostname_len = strlen(hostname);
+                size_t buffer_size = AWS_ARRAY_SIZE(socket->local_endpoint.address);
+                size_t to_copy = aws_min_size(hostname_len, buffer_size);
+                memcpy(socket->local_endpoint.address, hostname, to_copy);
+                socket->local_endpoint.port = port;
+            }
+            nw_release(local_endpoint);
```

Comment:
The local_endpoint doesn't appear to be released if there's no socket. We also appear to not use the hostname and port members if there's no socket. We could probably move hostname and port into this if block and move the nw_release of local_endpoint outside of it unless it's being taken care of elsewhere.